### PR TITLE
Array.Parallel.groupBy added

### PR DESCRIPTION
**Pull request by T-Gro**
This is next part of the [RFC](https://github.com/fsharp/fslang-design/blob/ff115e7f10993101c1624bfdc5dad9376d64553c/RFCs/1130-array-parallel-collection-functions-more-regular.md) to expand the Array.Parallel module, based on https://github.com/fsharp/fslang-suggestions/issues/187

[GroupBy Benchmarks](https://github.com/T-Gro/Fsharp.ParallelGroupBy.Benchmarks) were used for picking an appropriate implementation, comparing against:
Sequential Array.groupBy
PLINQ 

The tricky part was balancing the optimization with regards to varying settings. Beyond others:
Struct vs Ref
Expensive vs Cheap projection function
Cardinality of the results (# of buckets)

Some properties desired about the results:
- Invokes projection exactly once per input
- Has predictable allocation profile, allocating typically a lot less then PLINQ
- In most of the cases, at least 2x faster than LINQ (for some reasonable configurations, the speed over PLINQ can be up to 5x-7x better) 
- If it is slower then PLINQ, it is so only in rarer cases

One of the results contains measurement for the full configuration matrix:
https://github.com/T-Gro/Fsharp.ParallelGroupBy.Benchmarks/blob/main/Results/Detailed_Against_PLINQ_Full_Config_Matrix.md

### CODE REVIEW

**Review by KevinRansom — APPROVED** [review]

File: src/FSharp.Core/array.fs [hunk 1]
```diff
@@ -2054,6 +2056,126 @@ module Array =
...
+            |> ignore
+
+            finalResults
+
+        let groupByValueTypeParallel (keyf: 'T -> 'Key) (array: 'T[]) =
+            // Is it a bad idea to put floating points as keys for grouping? Yes
+            // But would the implementation fail with KeyNotFound "nan" if we just leave it? Also yes
+            // Here we  enforce nan=nan equality to prevent throwing
+            if typeof<'Key> = typeof<float> || typeof<'Key> = typeof<float32> then
+                let genericCmp =
```
Comment (psfinaki) [inline src/FSharp.Core/array.fs:2152]: What's cmp?

File: src/FSharp.Core/array.fs [hunk 2]
```diff
@@ -2054,6 +2056,126 @@ module Array =
...
+                        let idxToWrite = Interlocked.Decrement(counter)
+                        arrayForThisGroup[idxToWrite] <- array[elemIdx]
+            )
+            |> ignore
+
+            finalResults
+
+        let groupByValueTypeParallel (keyf: 'T -> 'Key) (array: 'T[]) =
+            // Is it a bad idea to put floating points as keys for grouping? Yes
+            // But would the implementation fail with KeyNotFound "nan" if we just leave it? Also yes
```
Comment (psfinaki) [inline src/FSharp.Core/array.fs:2149]: NaN

File: src/FSharp.Core/array.fs [hunk 3]
```diff
@@ -2054,6 +2056,126 @@ module Array =
...
+                        let (counter, arrayForThisGroup) = finalResultsLookup[key]
+                        let idxToWrite = Interlocked.Decrement(counter)
+                        arrayForThisGroup[idxToWrite] <- array[elemIdx]
+            )
+            |> ignore
+
+            finalResults
+
+        let groupByValueTypeParallel (keyf: 'T -> 'Key) (array: 'T[]) =
+            // Is it a bad idea to put floating points as keys for grouping? Yes
```
Comment (psfinaki) [inline src/FSharp.Core/array.fs:2148]: Why?

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/ArrayModule.fs [hunk 4]
```diff
@@ -1242,6 +1242,77 @@ type ArrayModule() =
         CheckThrowsArgumentNullException(fun () -> Array.groupBy funcInt (null : int array) |> ignore)
         ()
 
+    [<Fact>]
```
Comment (psfinaki) [inline tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/ArrayModule.fs:1245]: Honestly, the test is a bit hard to follow, I'd prefer having it split into parts (e.g. by each "compareAndAssert" statement)

File: src/FSharp.Core/array.fs [hunk 5]
```diff
@@ -2054,6 +2056,126 @@ module Array =
 
             result
 
+        // The following two parameters were benchmarked and found to be optimal.
```
Comment (psfinaki) [inline src/FSharp.Core/array.fs:2059]: Can benchmarks be committed as well?

**Review by psfinaki — APPROVED** [review]

File: src/FSharp.Core/array.fs (hunk 1 above)
Comment (T-Gro) [reply src/FSharp.Core/array.fs:2152]: shortcut for compare, as used in assemblers etc.

File: src/FSharp.Core/array.fs (hunk 2 above)
Comment (T-Gro) [reply src/FSharp.Core/array.fs:2149]: Why? even in F# it is called simply "nan" with that exact casing.

File: src/FSharp.Core/array.fs (hunk 3 above)
Comment (T-Gro) [reply src/FSharp.Core/array.fs:2148]: https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html 

Even naive looking numbers like 0.1 cannot be represented precisely.
Division by integers that are not powers of two will give a number that is only close to what a human expects.

If a business use case demands it, the right thing to do is to wrap them (e.g. via custom comparer, or a wrapper type) into something that will round the numbers to make them truly equal when needed.

File: src/FSharp.Core/array.fs (hunk 5 above)
Comment (T-Gro) [reply src/FSharp.Core/array.fs:2059]: There is a linked repo attached to this PR.
I do not think those belong to F# repo, they were just guidance for picking from a variety of implementation options.

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/ArrayModule.fs (hunk 4 above)
Comment (T-Gro) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/ArrayModule.fs:1245]: I chose consistency with this particular test suite (pretty much 1 function per collection function).
